lisa/sut_orchestrator/baremetal/platform_.py

```diff
@@ -0,0 +1,153 @@
+
+from lisa import RemoteNode, feature, schema
+from lisa.environment import Environment
+from lisa.platform_ import Platform
+from lisa.util import fields_to_dict
+from lisa.util.logger import Logger
+from lisa.util.shell import try_connect
+from lisa.util.subclasses import Factory
+
+from .. import BAREMETAL
+from . import features
+from .cluster.cluster import Cluster
+from .context import get_node_context
+from .ip_getter import IpGetterChecker
+from .readychecker import ReadyChecker
+from .schema import BareMetalPlatformSchema
+from .source import Source
+
+
+class BareMetalPlatform(Platform):
+    def __init__(
+        self,
+        runbook: schema.Platform,
+    ) -> None:
+        super().__init__(runbook=runbook)
+
+    @classmethod
+    def type_name(cls) -> str:
+        return BAREMETAL
+
+    @classmethod
+    def supported_features(cls) -> List[Type[feature.Feature]]:
+        return [features.StartStop, features.SerialConsole]
+
+    def _initialize(self, *args: Any, **kwargs: Any) -> None:
+        baremetal_runbook: BareMetalPlatformSchema = self.runbook.get_extended_runbook(
+            BareMetalPlatformSchema
+        )
+        assert baremetal_runbook, "platform runbook cannot be empty"
+        self._baremetal_runbook = baremetal_runbook
+        self.source_path: str = ""
+        self.copy_build: bool = False
+        self.ready_checker_factory = Factory[ReadyChecker](ReadyChecker)
+        self.cluster_factory = Factory[Cluster](Cluster)
+        self.ip_getter_factory = Factory[IpGetterChecker](IpGetterChecker)
+        self.source_factory = Factory[Source](Source)
+
+    def _prepare_environment(self, environment: Environment, log: Logger) -> bool:
+        return True
+
+    def _deploy_environment(self, environment: Environment, log: Logger) -> None:
+        # currently only support one cluster
+        assert self._baremetal_runbook.cluster, "no cluster is specified in the runbook"
+        cluster_instance = self._baremetal_runbook.cluster[0]
+
+        cluster = self.cluster_factory.create_by_runbook(cluster_instance)
+        assert cluster.runbook.client, "no client is specified in the runbook"
+
+        # copy build (shared, check if it's copied)
+        if self._baremetal_runbook.source and not self.source_path:
+            source = self.source_factory.create_by_runbook(
+                self._baremetal_runbook.source
+            )
+            self._log.debug(f"found build '{source.type_name()}', to expand runbook.")
+            self.source_path = source.download()
+
+        # ready checker cleanup
+        if cluster_instance.ready_checker:
+            ready_checker = self.ready_checker_factory.create_by_runbook(
+                cluster_instance.ready_checker
+            )
+            ready_checker.clean_up()
+
+        # copy build if source exists
+        if not self.copy_build:
+            cluster.copy(source_path=self.source_path)
+            self.copy_build = True
```

Comment:
This flag shouldn't be in platform level, it should mark a build is copied to a cluster. if it's in platform and there are two clusters, the build is not able to be copied on second cluster.

Reply:
updated, not test yet